### PR TITLE
feat(comments): added threaded comments capabilities

### PR DESCRIPTION
**Pull request by jdalsem**
- [x] add level = 1 migration
 - [x] write docs? or migration appendix?
 - [x] unit tests (elggcomment)
 - [x] integration tests (action comment/save)
 - [x] integration test for thread preloader

### CODE REVIEW

File: engine/classes/Elgg/Comments/Preloader.php [hunk 1]
```diff
@@ -114,7 +114,7 @@ public static function preload(\Elgg\Hook $hook): void {
 		if (!$preload) {
 			return;
 		}
-
```
Comment (jeabakker) [inline engine/classes/Elgg/Comments/Preloader.php:117]: shame, if this is the only change please undo this to the file doesn't get touched

File: engine/classes/ElggComment.php [hunk 2]
```diff
@@ -15,6 +19,55 @@ protected function initializeAttributes() {
...
 		$this->attributes['subtype'] = 'comment';
+		
+		$this->level = 1;
+	}
+	
+	/**
+	 * {@inheritDoc}
+	 */
+	public function delete($recursive = true) {
+		$result = parent::delete();
```
Comment (jeabakker) [inline engine/classes/ElggComment.php:30]: pass `$recursive` please

File: engine/classes/ElggComment.php [hunk 3]
```diff
@@ -15,6 +19,55 @@ protected function initializeAttributes() {
...
+					'batch_inc_offset' => false,
+					'metadata_name_value_pairs' => [
+						'name' => 'parent_guid',
+						'value' => $this->guid,
+					],
+				]);
+				
+				/* @var $child \ElggComment */
+				foreach ($children as $child) {
+					$child->delete();
```
Comment (jeabakker) [inline engine/classes/ElggComment.php:49]: pass `$recursive` please

File: actions/comment/save.php [hunk 4]
```diff
@@ -39,26 +39,37 @@
...
 	}
 
-	$user = elgg_get_logged_in_user_entity();
-
 	$comment = new ElggComment();
 	$comment->description = $comment_text;
-	$comment->owner_guid = $user->getGUID();
-	$comment->container_guid = $entity->getGUID();
-	$comment->access_id = $entity->access_id;
+	$comment->owner_guid = elgg_get_logged_in_user_guid();
```
Comment (jeabakker) [inline actions/comment/save.php:44]: isn't `owner_guid` defaulted to the logged in user?

File: engine/schema/migrations/20211224075815_set_comments_level_one.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+declare(strict_types=1);
+
+use Phinx\Migration\AbstractMigration;
+
+final class SetCommentsLevelOne extends AbstractMigration {
+    /**
+     * Add level = 1 metadata to comments
+     */
+    public function up() {
+    	$prefix = $this->getAdapter()->getOption('table_prefix');
```
Comment (jeabakker) [inline engine/schema/migrations/20211224075815_set_comments_level_one.php:11]: indentation

File: engine/tests/phpunit/integration/Elgg/Actions/Comment/SaveIntegrationTest.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		// validate created sub comment entity
+		$comments = elgg_get_entities(['type' => 'object', 'subtype' => 'comment', 'metadata_name_value_pairs' => ['parent_guid' => $top_comment->guid]]);
+		$this->assertCount(1, $comments);
+		
+		$sub_comment = $comments[0];
+		$this->assertEquals('bar', $sub_comment->description);
+		$this->assertEquals(2, $sub_comment->level);
+		$this->assertEquals(2, $sub_comment->getLevel());
+		$this->assertEquals($top_comment->guid, $sub_comment->parent_guid);
+		$this->assertEquals($top_comment->guid, $sub_comment->thread_guid);
```
Comment (jeabakker) [inline engine/tests/phpunit/integration/Elgg/Actions/Comment/SaveIntegrationTest.php:55]: also assert that the subcomment has the original `$entity` as it's container

File: engine/tests/phpunit/integration/Elgg/Actions/Comment/SaveIntegrationTest.php [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+		// validate created sub comment entity
+		$comments = elgg_get_entities(['type' => 'object', 'subtype' => 'comment', 'metadata_name_value_pairs' => ['parent_guid' => $sub_comment->guid]]);
+		$this->assertCount(1, $comments);
+		
+		$lowest_comment = $comments[0];
+		$this->assertEquals('foobar', $lowest_comment->description);
+		$this->assertEquals(3, $lowest_comment->level);
+		$this->assertEquals(3, $lowest_comment->getLevel());
+		$this->assertEquals($sub_comment->guid, $lowest_comment->parent_guid);
+		$this->assertEquals($top_comment->guid, $lowest_comment->thread_guid);
```
Comment (jeabakker) [inline engine/tests/phpunit/integration/Elgg/Actions/Comment/SaveIntegrationTest.php:74]: also assert that the threaded comment has the original `$entity` as it's container

File: engine/tests/phpunit/unit/ElggCommentUnitTest.php [hunk 8]
```diff
@@ -1,11 +1,16 @@
 <?php
-
```
Comment (jeabakker) [inline engine/tests/phpunit/unit/ElggCommentUnitTest.php:2]: no, this could confuse the comment below between file comment and class comment

Comment (jeabakker) [reply engine/tests/phpunit/unit/ElggCommentUnitTest.php:2]: Or also remove the class comment

File: views/default/object/comment.php [hunk 9]
```diff
@@ -33,19 +33,33 @@
...
 		'access' => false,
 		'title' => false,
+		'show_summary' => true,
 		'content' => $body,
 		'imprint' => elgg_extract('imprint', $vars, []),
 		'class' => elgg_extract_class($vars),
 	];
 	$params = $params + $vars;
 	
+	if (!empty(elgg()->thread_preloader->getChildren($comment->guid))) {
```
Comment (jeabakker) [inline views/default/object/comment.php:59]: wouldn't it be nicer if the preloader has a function `hasChildren()`

File: views/default/page/elements/comments.php [hunk 10]
```diff
@@ -38,30 +38,60 @@
...
+	
+	$options['metadata_name_value_pairs'] = ['level' => 1];
+	
+	$show_guid = (int) elgg_extract('show_guid', $vars);
+	if ($show_guid && $limit) {
+		// show the offset that includes the comment
+		$operator = $latest_first ? '>' : '<';
+		$condition = function(QueryBuilder $qb) use ($show_guid, $operator) {
+			return $qb->compare('e.guid', $operator, $show_guid, ELGG_VALUE_INTEGER);
+		};
```
Comment (jeabakker) [inline views/default/page/elements/comments.php:63]: if this is only used once just add it to the `wheres` directly

**Review by jeabakker — CHANGES_REQUESTED** [review]

File: views/default/object/comment.php (hunk 9 above)
Comment (jdalsem) [reply views/default/object/comment.php:59]: it is nicer, but for now i will not add it as it is Friday

File: views/default/page/elements/comments.php (hunk 10 above)
Comment (jdalsem) [reply views/default/page/elements/comments.php:63]: 👍

**Review by jeabakker — APPROVED** [review]

**Pull request merged**